### PR TITLE
JENKINS-39856 fixed memory leak caused by jsplumb and isolation of pipeline function to enable multiple dashboard DPP portlet

### PR DESCRIPTION
**Pull request by pskumar448**

### CODE REVIEW

File: src/main/webapp/pipe.js [hunk 1]
```diff
@@ -1,31 +1,31 @@
 function pipelineUtils() {
-     self = this;
-     this.updatePipelines = function(divNames, errorDiv, view, fullscreen, page, component, showChanges, aggregatedChangesGroupingPattern, timeout, pipelineid) {
+     var self = this;
+     this.updatePipelines = function(divNames, errorDiv, view, fullscreen, page, component, showChanges, aggregatedChangesGroupingPattern, timeout, pipelineid, jsplumb) {
```
Comment (tommysdk) [inline src/main/webapp/pipe.js]: do we need to pass the jsplumb here? is it to satisfy the portlet setup?

**Comment by tommysdk** [comment]
@pskumar448 I'll give this PR a spin on Monday!

File: src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly [hunk 2]
```diff
@@ -5,18 +5,21 @@
     <st:bind var="view" value="${it}"/>
     <st:bind var="fullscreen" value='${request.getParameter("fullscreen")}'/>
 
-    <j:set var="fullscreen" value='${request.getParameter("fullscreen") == null ? "false" : request.getParameter("fullscreen")}'/>
+    <j:set var="fullscreen"
```
Comment (pskumar448) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly]: This whitespace occurred while reformatting code in IDE

**Comment by pskumar448** [comment]
@tommysdk 
This fix related to memory leak issue but it extended to make it work for multiple portlets.
Let me know shall I create a new JIRA ticket for this.

**Comment by pskumar448** [comment]
@tommysdk 
Update me with your feedback, once you have verified memory leak issue.

**Comment by tommysdk** [comment]
@pskumar448 absolutely! I am evaluating the the proposed solution on a long running Jenkins instance. Will evaluate the data tomorrow and let you know!

File: src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly [hunk 3]
```diff
@@ -5,18 +5,21 @@
...
+           value='${request.getParameter("fullscreen") == null ? "false" : request.getParameter("fullscreen")}'/>
 
     <st:bind var="page" value='${request.getParameter("page")}'/>
     <j:set var="page" value='${request.getParameter("page") == null ? 1 : request.getParameter("page")}'/>
 
     <st:bind var="component" value='${request.getParameter("component")}'/>
-    <j:set var="component" value='${request.getParameter("component") == null ? 1 : request.getParameter("component")}'/>
+    <j:set var="component"
```
Comment (tommysdk) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly]: same as above

File: src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly [hunk 4]
```diff
@@ -5,18 +5,21 @@
...
+           value='${request.getParameter("component") == null ? 1 : request.getParameter("component")}'/>
 
     <link rel="stylesheet" href="${resURL}/plugin/delivery-pipeline-plugin/themes/${it.theme}/pipeline-common.css"/>
     <j:switch on="${fullscreen}">
         <j:case value="true">
-            <link rel="stylesheet" href="${resURL}/plugin/delivery-pipeline-plugin/themes/${it.theme}/pipeline-fullscreen.css"/>
+            <link rel="stylesheet"
```
Comment (tommysdk) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly]: dito

File: src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly [hunk 5]
```diff
@@ -57,41 +61,49 @@
             </div>
 
             <div id="pipelineerror-${it.id == null ? 0 : it.id}" class="pipelineerror"/>
-            <div id="pipeline-message-${it.id == null ? 0 : it.id}" class="pipeline-message"><div class="pipeline-loading-icon"></div></div>
+            <div id="pipeline-message-${it.id == null ? 0 : it.id}" class="pipeline-message">
+                <div class="pipeline-loading-icon"></div>
+            </div>
 
             <j:forEach begin="1" end="${it.noOfColumns}" indexVar="i">
-                <div id="pipelines-${i}-${it.id == null ? 0 : it.id}" class="left" style="width: ${100/it.noOfColumns}%;"></div>
```
Comment (tommysdk) [inline src/main/resources/se/diabol/jenkins/pipeline/DeliveryPipelineView/main.jelly]: here as well (?)

File: src/main/webapp/pipe.js (hunk 1 above)
Comment (tommysdk) [inline src/main/webapp/pipe.js]: Got it. The solution looks sufficient

**Comment by tommysdk** [comment]
I've assessed this feature request and the functionality is deemed sufficient. The memory leak in 0.10.0 is addressed and if this fixes the broken dashboard portlet we can merge this. Before merge I would however want to consider adding a reproducing test case and address the minor code comments I added.

**Comment by pskumar448** [comment]
@tommysdk 
Addressed the code formatting issues, looking into adding a test case for this scenario. Is that possible to give me information like whether to add JavaScript test or GUI test.

-Suresh

**Comment by patbos** [comment]
The integration tests are failing, please run
`mvn integration-test`

There is a couple of Selenium tests in 
https://github.com/Diabol/delivery-pipeline-plugin/blob/master/src/test/java/se/diabol/jenkins/pipeline/functionaltest/GuiFunctionalIT.java

**Comment by pskumar448** [comment]
@patbos 
Fixed failed functional tests.

**Comment by pskumar448** [comment]
@tommysdk 
Added Gui tests to check the fix for memory leak.
Let me know your feedback regarding this test cases added

**Comment by tommysdk** [comment]
@pskumar448 awesome! I'll take a look Friday!

**Comment by tommysdk** [comment]
Added commit and pushed as [PR 222](https://github.com/Diabol/delivery-pipeline-plugin/pull/222).

**Comment by tommysdk** [comment]
Merged as [PR 222](https://github.com/Diabol/delivery-pipeline-plugin/pull/221). Thank you @pskumar448 for yet another contribution!

**Pull request closed without merging**